src/EigenvalueSolverTypes.f90

```diff
@@ -253,6 +253,52 @@ SUBROUTINE evsolver_sub_absintfc(solver)
   CONTAINS
 !
 !-------------------------------------------------------------------------------
+!> @brief Create an Eigenvalue solver object compatible with the provided matrix
+!>
+!> @param solver a pointer to the base eigenvalue solver type. Should be NULL
+!> @param MPIEnv the MPI environment to be used to initialize the solver
+!> @param M the matrix to use to determine which type of solver to create
+!> @param params A parameter list to use in allocating the solver
+!> 
+!> This is a simple abstract factory routine for the base EigenvalueSolverType.
+!> It uses the type of the passed MatrixType object to determine which type of
+!> eigenvalue solver to allocate, then initialized the solver with the passed
+!> parameters. The purpose of such a routine is to hide from the client code the
+!> actual type of the matrix/solver pair, and the preprocessor guards necessary
+!> to safely use the TPLs.
+    SUBROUTINE EigenvalueSolverFactory(solver,MPIEnv,M,params)
+      CHARACTER(LEN=*),PARAMETER :: myName="EigenvalueSolverFactory"
+      CLASS(EigenvalueSolverType_Base),INTENT(INOUT),POINTER :: solver
+      TYPE(MPI_EnvType),INTENT(IN),TARGET :: MPIEnv
+      CLASS(MatrixType),INTENT(IN),TARGET :: M
+      TYPE(ParamType),INTENT(IN) :: params
+      
+      IF(ASSOCIATED(solver)) THEN
+        CALL eEigenvalueSolverType%raiseError(modName//"::"//myName//" - "// &
+          "Solver pointer is already associated")
+        RETURN
+      ENDIF
+
+      SELECTTYPE(M)
+#ifdef FUTILITY_HAVE_PETSC
```

Comment:
What was the reasoning for ifdefing the TPL provided types? (thinking about this from the point of view of minimizing number places with ifdefs)

I guess you get build errors instead of run time errors if you still provide stub types. That seems like an advantage. Also, I guess you should only need the ifdefs in the factories and modules containing the concrete implementations.